test/samplers/test_samplers.py

```diff
@@ -0,0 +1,202 @@
+    # seed that works.
+    torch.manual_seed(0)
+
+    decoder = SimpleVideoDecoder(NASA_VIDEO.path)
+
+    clips = clips_at_random_indices(
+        decoder,
+        num_clips=10,
+        num_frames_per_clip=2,
+        sampling_range_start=sampling_range_start,
+        sampling_range_end=sampling_range_end,
+    )
+
+    cm = (
+        contextlib.nullcontext()
+        if assert_all_equal
+        else pytest.raises(AssertionError, match="Tensor-likes are not")
+    )
+    with cm:
+        for clip in clips:
+            assert_tensor_equal(clip.data, clips[0].data)
+
+
+def test_random_sampler_range_negative():
+    # Test the passing negative values for sampling_range_start and
+    # sampling_range_end is the same as passing `len(decoder) - val`
+
+    decoder = SimpleVideoDecoder(NASA_VIDEO.path)
+
+    clips_1 = clips_at_random_indices(
+        decoder,
+        num_clips=10,
+        num_frames_per_clip=2,
+        sampling_range_start=len(decoder) - 100,
+        sampling_range_end=len(decoder) - 99,
+    )
+
+    clips_2 = clips_at_random_indices(
+        decoder,
+        num_clips=10,
+        num_frames_per_clip=2,
+        sampling_range_start=-100,
+        sampling_range_end=-99,
+    )
+
+    # There is only one unique clip in clips_1...
+    for clip in clips_1:
+        assert_tensor_equal(clip.data, clips_1[0].data)
+    # ... and it's the same that's in clips_2
+    for clip in clips_2:
+        assert_tensor_equal(clip.data, clips_1[0].data)
+
+
+def test_random_sampler_randomness():
+    decoder = SimpleVideoDecoder(NASA_VIDEO.path)
+    num_clips = 5
+
+    builtin_random_state_start = random.getstate()
+
+    torch.manual_seed(0)
+    clips_1 = clips_at_random_indices(decoder, num_clips=num_clips)
+
+    # Assert the clip starts aren't sorted, to make sure we haven't messed up
+    # the implementation. (This may fail if we're unlucky, but we hard-coded a
+    # seed, so it will always pass.)
+    clip_starts = [clip.pts_seconds.item() for clip in clips_1]
+    assert sorted(clip_starts) != clip_starts
+
+    # Call the same sampler again with the same seed, expect same results
+    torch.manual_seed(0)
+    clips_2 = clips_at_random_indices(decoder, num_clips=num_clips)
+    for clip_1, clip_2 in zip(clips_1, clips_2):
+        assert_tensor_equal(clip_1.data, clip_2.data)
+        assert_tensor_equal(clip_1.pts_seconds, clip_2.pts_seconds)
+        assert_tensor_equal(clip_1.duration_seconds, clip_2.duration_seconds)
+
+    # Call with a different seed, expect different results
+    torch.manual_seed(1)
+    clips_3 = clips_at_random_indices(decoder, num_clips=num_clips)
+    with pytest.raises(AssertionError, match="Tensor-likes are not"):
+        assert_tensor_equal(clips_1[0].data, clips_3[0].data)
+
+    # Make sure we didn't alter the builtin Python RNG
+    builtin_random_state_end = random.getstate()
+    assert builtin_random_state_start == builtin_random_state_end
+
```

Comment:
I wonder if we can fuzz test this to make sure we don't return errors mid-training.

Can pytest help with passing in random values to the parameters (with certain constraints like positive-only values, etc. for things like frames_per_clip, etc.) to make sure we are robust to errors?

Reply:
There is the `hypothesis` package that I think does what you're suggesting, but I'm not sure how useful it would be in this specific instance? What do you have in mind with 

> make sure we don't return errors mid-training.

?

Reply:
I was requesting fuzz testing so we can be robust to errors so we don't return errors in the middle of training (which is what your original philosophy was).

<img width="773" alt="image" src="https://github.com/user-attachments/assets/de62305e-d380-4770-a2ba-70c3c8ef4857">

Now if we are switching to errors mindset, you do seem to have tests for exact errors, so you can resolve this comment.